include/realizations/catchment/Bmi_Cpp_Adapter.hpp

```diff
@@ -0,0 +1,473 @@
+            const std::string
+            get_analogous_cxx_type(const std::string &external_type_name, const size_t item_size) override {
+                return external_type_name;
+            }
+
+            /**
+             * Whether the backing model has been initialized yet.
+             *
+             * @return Whether the backing model has been initialized yet.
+             */
+            bool is_model_initialized();
+
+            void SetValue(std::string name, void *src) override;
+
+            template<class T>
+            void SetValue(std::string name, std::vector<T> src) {
+                bmi_model->SetValue(name, src);
+            }
+
+            void SetValueAtIndices(std::string name, int *inds, int count, void *src) override;
+
+            template<class T>
+            void SetValueAtIndices(const std::string &name, std::vector<int> inds, std::vector<T> src) {
+                bmi_model->SetValueAtIndices(name, inds, src);
+            }
+
+            void Update() override;
+
+            void UpdateUntil(double time) override;
+
+        protected:
+            // TODO: look at setting this in some other way
+            std::string model_name = "BMI C++ model";
+
+            /**
+             * Construct the backing BMI model object, then call its BMI-native ``Initialize()`` function.
+             *
+             * Implementations should return immediately without taking any further action if ``model_initialized`` is
+             * already ``true``.
+             *
+             * The call to the BMI native ``Initialize(string)`` should pass the value stored in ``bmi_init_config``.
+             * @see construct_and_init_backing_model_for_type
+             */
+            void construct_and_init_backing_model() override {
+                construct_and_init_backing_model_for_type();
+            }
+
+            /**
+             * Load and execute the creator function for the backing BMI module.
+             *
+             * Integrated BMI module libraries are expected to provide an additional ``bmi_model_create`` function. This
+             * works like a constructor (or factory) for the model object and returns a pointer to the allocated model
+             * instance.
+             */
+            inline void execModuleCreation() {
+                if (get_dyn_lib_handle() == nullptr) {
+                    dynamic_library_load();
+                }
+                void *symbol;
+                //Cpp_Bmi *(*dynamic_create_bmi)(Cpp_Bmi *model);
+                ModelCreator dynamic_creator;
+
+                try {
+                    // Acquire the BMI creator function
+                    symbol = dynamic_load_symbol(model_create_fname);
+                    dynamic_creator = (ModelCreator) symbol;
+                    bmi_model = std::shared_ptr<Cpp_Bmi>(
+                        dynamic_creator(),
+                        [](Cpp_Bmi* p) {
+                            // DO NOTHING.
```

Comment:
Looking for feedback on the below. It may make the most sense to simply go back to classic pointers... this is what I evolved to. ???

Reply:
Okay, I may be replying too quick here, but... Yes, that's correct... my assumption was that we treat every adapter as if it were a plain BMI C++ model... that means that there must be a step # 2.(n-1) where `Finalize()` is called on the adapter (which _should_ be proxied to the underlying model object in the adapter's `Finalize()` method, but was missed above, per my comment there). Are we not calling `Finalize()` on the adapters before deleting them in step 3? If not... I might start to argue that we should be. Otherwise we're implementing the BMI interface according to the .hpp file but are not using the interface according to its semantics (I'm assuming those semantics require or imply that `Finalize()` should be called before `delete`.

Reply:
A few things:

- `finalizeForLibAbstraction()` is deliberately not virtual, so the logic needed for `Finalize()` could be called from the destructor more safely (and then reused more easily by `Finalize()` itself)
- See earlier comment again about how this type's destructor probably needs to be calling `Finalize()` or something equivalent
- I agree with the earlier comment, and the one above, that the module's own `Finalize()` doesn't seem to be getting called as part of this object's `Finalize()`, but should be
- Yes, we are implementing the BMI interface, but we are not limiting ourselves to exclusively using the semantics of the interface 
  - It makes sense to have and adapter implement the interface, as this aligns it intuitively for its role as a proxy
  - It should not be necessary to restrict a BMI implementation to only work (internally) via the BMI interface functions
  - It doesn't make sense to restrict the framework to _only_ working with an adapter via BMI, because we have the code (e.g., why require explicitly calling `Finalize()` when it makes sense to have this happen during the destructor?)
  - However, adapters _must_ be written to work in a consistent manner whether we follow the more strict BMI workflow (e.g., explicitly call `Finalize()` somewhere) or not (e.g., expect the destructor to also handling `Finalize()`)

Reply:
> * It doesn't make sense to restrict the framework to _only_ working with an adapter via BMI, because we have the code (e.g., why require explicitly calling `Finalize()` when it makes sense to have this happen during the destructor?)

Well, I'd argue It does make sense to restrict ourselves, at least (or especially) _if_ we ever expected to treat adapters and non-adapter C++ BMI models interchangeably...for which we might _not_ have the code. It's not an explicit part of the interface, but is probably implicit in the name--that if you don't call it explicitly then memory may not be freed properly... at the least, we couldn't assume that the non-adapter model would call Finalize on itself in its destructor. It becomes a LSP thing at that point.

A related question...are we considering it safe to call `Finalize()` more than once? I would generally think we'd assume that was not safe, but I think what's being discussed makes that possible or even likely. Again, the BMI docs are not helpful here.

Reply:
Back on this... repeating the question if we're assuming it's safe to call `Finalize()` on a BMI model more than once? Or is this a restriction we want to place on models in the framework, that they ensure that it's safe to do so? Sounds like something that would be likely to segfault if the model developer is not expecting it to happen.

Reply:
@mattw-nws, FWIW, we are actually getting the underlying C struct and Python model in certain tests.   It may still be more appropriate to change to `unique_ptr` and do something differently in the tests, but that'd be another aspect to any changes.

Reply:
Refactoring the bmi_model property seems out of scope. Other topics surrounding `Finalize()` are resolved in [d69094a](https://github.com/NOAA-OWP/ngen/pull/343/commits/d69094a1f9606e5d11cecdfa865601cc9218938a). Resolve this conversation, @hellkite500 and @robertbartel ?

Reply:
Did you mean this commit? [d73eee3](https://github.com/NOAA-OWP/ngen/pull/343/commits/d73eee339d2ef1a149c6cadcf1fb265eea8c6e25)

Reply:
@hellkite500 yes